runtime/compiler/z/codegen/J9TreeEvaluator.cpp

```diff
@@ -1961,6 +1961,376 @@ J9::Z::TreeEvaluator::inlineUTF16BEEncode(TR::Node *node, TR::CodeGenerator *cg)
+   TR::Register* arr = cg->gprClobberEvaluate(node->getChild(1));
+   TR::Register* off = cg->gprClobberEvaluate(node->getChild(2));
+   TR::Register* end = cg->gprClobberEvaluate(node->getChild(3));
+
+   // Registers
+   TR::Register* tablePtr = cg->allocateRegister();
+   TR::Register* buf = cg->allocateRegister();
+   TR::Register* len = cg->allocateRegister();
+
+   // Zero extend incoming 32 bit values
+   generateRREInstruction(cg, TR::InstOpCode::LLGFR, node, off, off);
+   generateRREInstruction(cg, TR::InstOpCode::LLGFR, node, len, end);
+
+   // Adjust data pointer and length for offset index
+   generateRXInstruction(cg, TR::InstOpCode::getLoadAddressOpCode(), node, buf, generateS390MemoryReference(arr, off, isRawAddress ? 0 : TR::Compiler->om.contiguousArrayHeaderSizeInBytes(), cg));
+   generateRRInstruction(cg, TR::InstOpCode::getSubstractRegOpCode(), node, len, off);
+
+   // Vector constants
+   TR::Register* vConstPermLE2BE = cg->allocateRegister(TR_VRF);
+   TR::Register* vConstR2R1 = cg->allocateRegister(TR_VRF);
+   TR::Register* vConstR4R3 = cg->allocateRegister(TR_VRF);
+   TR::Register* vConstR5 = cg->allocateRegister(TR_VRF);
+   TR::Register* vConstRUPoly = cg->allocateRegister(TR_VRF);
+   TR::Register* vConstCRCPoly = cg->allocateRegister(TR_VRF);
+
+   // Vectors for CRC computation
+   TR::Register* vFold1 = cg->allocateRegister(TR_VRF);
+   TR::Register* vFold2 = cg->allocateRegister(TR_VRF);
+   TR::Register* vFold3 = cg->allocateRegister(TR_VRF);
+   TR::Register* vFold4 = cg->allocateRegister(TR_VRF);
+   TR::Register* vShift = vConstPermLE2BE;
+   TR::Register* vScratch = cg->allocateRegister(TR_VRF);
+
+   // Next data chunks that are fetched from the input buffer
+   TR::Register* vInput1 = cg->allocateRegister(TR_VRF);
+   TR::Register* vInput2 = cg->allocateRegister(TR_VRF);
+   TR::Register* vInput3 = cg->allocateRegister(TR_VRF);
+   TR::Register* vInput4 = cg->allocateRegister(TR_VRF);
+
+   // Create the necessary labels
+   TR::LabelSymbol* startICF = generateLabelSymbol(cg);
+   TR::LabelSymbol* foldBy4Header = generateLabelSymbol(cg);
+   TR::LabelSymbol* foldBy4Loop = generateLabelSymbol(cg);
+   TR::LabelSymbol* foldby4LoopEntry = generateLabelSymbol(cg);
+   TR::LabelSymbol* reduce4to1 = generateLabelSymbol(cg);
+   TR::LabelSymbol* foldBy1Header = generateLabelSymbol(cg);
+   TR::LabelSymbol* foldBy1Loop = generateLabelSymbol(cg);
+   TR::LabelSymbol* foldBy1LoopEntry = generateLabelSymbol(cg);
+   TR::LabelSymbol* finalReduction = generateLabelSymbol(cg);
+   TR::LabelSymbol* callJava = generateLabelSymbol(cg);
+   TR::LabelSymbol* endICF = generateLabelSymbol(cg);
+
+   // For debug annototations
+   TR::Instruction *cursor;
+   TR_Debug *debugObj = cg->getDebug();
+
+   // If less than 16B of input, branch to bytewise path
+   // The vectorized path only works for multiples of 16B
+   generateRIInstruction(cg, TR::InstOpCode::CGHI, node, len, 16);
```

Comment:
We should be able to use `generateS390CompareAndBranchInstruction` here. It would generate compare and branch instruction on supported platform [1]. We should have plenty of examples where this is used. You can check them out. 

[1]. https://github.com/eclipse/omr/blob/a9329039e91e2a88de7bc91a866d79e7bfdd4993/compiler/z/codegen/S390GenerateInstructions.cpp#L381